### PR TITLE
Quick local play using KIND

### PR DESCRIPTION
**Pull request by eladiw**
**Tl;dr**
Clone the repo, Navigate to the 'deployment' directory and run './run-with-kind.sh'.

**Details:**
KIND's prerequisites are Docker (must) and Go (Optional).
By actually running make build we can disregard the Go prerequisites (Go will be fetched while building).

**How to test:**
Option 1 - Mac / Linux users. run as documented.

Option 2 - Windows / Mac / Linux can follow these steps:

0. Create an Ubuntu VM (make sure you have at least 4 CPUs, or the analyzer won't start and make sure the machine as at least 8 GB ram).

1. install make:

```sh
sudo apt install make
```

2. Install [docker](https://docs.docker.com/install/linux/docker-ce/ubuntu/)
```sh
sudo apt-get update

sudo apt-get install \
    apt-transport-https \
    ca-certificates \
    curl \
    gnupg-agent \
    software-properties-common

curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -

sudo add-apt-repository \
   "deb [arch=amd64] https://download.docker.com/linux/ubuntu \
   $(lsb_release -cs) \
   stable"

sudo apt-get update

sudo apt-get install docker-ce docker-ce-cli containerd.io

```
3. Verify Docker was installed :

```sh
sudo docker run hello-world
```

4. Install kubectl:

```sh
curl -LO https://storage.googleapis.com/kubernetes-release/release/`curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt`/bin/linux/amd64/kubectl

chmod +x ./kubectl

sudo mv ./kubectl /usr/local/bin/kubectl

kubectl version --client
```

5. Install Helm:
```sh
curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3

chmod 700 get_helm.sh

./get_helm.sh
```

Now you have Docker, kubectl and Helm.
6. Install Presidio:
```sh
cd deployment
./run-with-kind.sh
```
Wait for Presidio to finish starting up (all pods running)
port forward to the api micro service, give it a try
for example:

**create template:**

```sh
curl -vL -H "Content-Type: application/json" -X POST -d '{"fields": [{ "name": "PASSPORT"}, { "name": "PERSON" }]}' http://localhost:8080/api/v1/templates/123/analyze/1234
```

**Analyze:**

```sh
curl -vL -H "Content-Type: application/json" -X POST -d '{ "text": "Hi John, here is X4316545 please make update it today in Israel","analyzeTemplateId": "1234"}' http://localhost:8080/api/v1/projects/123/analyze
```

### CODE REVIEW

File: README.MD [hunk 1]
```diff
@@ -88,6 +88,28 @@ Presidio leverages:
 * [GRPC](https://grpc.io)
 * [re2](https://github.com/google/re2)
 
+## Give it a try - Run locally with KIND
+Presidio is built for Kubernetes, however you can give it a try using [KIND (Kubernetes IN Docker)](https://github.com/kubernetes-sigs/kind).
```
Comment (balteravishay) [inline README.MD]: KIND is kubernetes. "However" doesn't belong here

File: deployment/deploy-presidio.sh [hunk 2]
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
 REGISTRY=${1:-mcr.microsoft.com}
 TAG=${2:-latest}
-helm install --name redis stable/redis --set usePassword=false,rbac.create=true --namespace presidio-system --wait
-helm install --name presidio-demo --set registry=$REGISTRY,tag=$TAG ../charts/presidio --namespace presidio
+helm install redis stable/redis --set usePassword=false,rbac.create=true --namespace presidio-system --wait
```
Comment (balteravishay) [inline deployment/deploy-presidio.sh]: perhaps add in docs that we support now only helm3 (which is why this was changed)

File: deployment/run-with-kind.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+# This script utilize KIND (https://github.com/kubernetes-sigs/kind), to install Presidio
+# into docker (with k8s emulation).
+# This script is intended to get Presidio up and running in no time, so you can play with it.
+# However it is NOT intended to be installed in this manner for any workload (dev or prod).
+
+# Clone KIND and build it (this way we eliminate the need to have GO installed)
```
Comment (balteravishay) [inline deployment/run-with-kind.sh]: not sure this is the best approach over having kind already installed. keeping this to see what other think.
downside is that we build kind every time.

**Comment by uptowndat777777** [comment]
Hello, Compliment of the day, how are you today ? i am Happy to get together with you,My name is Serach , I'm a military medical nurse from israel. Please let not my profession, race or nationality inch your mind, I have gone across your profile and became interested in you, we can be good friends, contact me through my private email address/  //abelmans@aol.com   ////

File: deployment/run-with-kind.sh (hunk 3 above)
Comment (eladiw) [reply deployment/run-with-kind.sh]: I'm thinking about it from a new user perspective. 
How many of those potential users have Docker and how many have Kind.
If you have kind, you also have the presidio chart...just install it.

otherwise, we can automate it.

File: README.MD [hunk 4]
```diff
@@ -88,6 +88,28 @@ Presidio leverages:
...
+
+1. Install [Docker](https://docs.docker.com/install/).
+2. Clone Presidio.
+3. Run
+
+    ```sh
+    cd deployment/
+    ./run-with-kind.sh
```
Comment (omri374) [inline README.MD]: Some issues I had with installing on an ubuntu VM:
1. ./run-with-kind.sh required sudo
2. After it finishes, it says "presidio is now installed! To find out about your newly configured system, run: 
`$ help status presidio-demo`
However running this (with sudo) returns:
`Error: release: not found`
3. Port forwarding required sudo in both commands:
`sudo kubectl port-forward -n presidio $(sudo kubectl get pod -n presidio -l app=presidio-demo-presidio-api -o jsonpath='{.items[0].metadata.name}') 8080:8080`
Worth adding an & at the end as it blocks the terminal window

Comment (omri374) [reply README.MD]: After completing this, all worked well. Pretty cool!
Maybe we should have the prerequisites scripted in prerequisites.sh (and say that if you don't need to install a b c you can skip it). The other script should install kind and presidio (and maybe do port forwarding?)
I agree we shouldn't assume that users have kind installed.

**Comment by itye-msft** [comment]
```
kubectl port-forward -n presidio $(kubectl get pod -n presidio -l app=presidio-demo-presidio-api -o jsonpath='{.items[0].metadata.name}') 8080:8080
Forwarding from 127.0.0.1:8080 -> 8080
Forwarding from [::1]:8080 -> 8080
Handling connection for 8080
E0311 16:57:39.830733   39428 portforward.go:400] an error occurred forwarding 8080 -> 8080: error forwarding port 8080 to pod 2fd642eef7855fb1210349f1c1d859bdd13b5e8af1c6d99265c7e401eca119f3, uid : failed to execute portforward in network namespace "/var/run/netns/cni-ac2f8fad-bb28-0060-11bf-0480eef50916": socat command returns error: exit status 1, stderr: "2020/03/11 14:57:39 socat[3007] E connect(5, AF=2 127.0.0.1:8080, 16): Connection refused\n"
```

```
curl -vL -H "Content-Type: application/json" -X POST -d '{"fields": [{ "name": "PASSPORT"}, { "name": "PERSON" }]}' http://localhost:8080/api/v1/templates/123/analyze/1234
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> POST /api/v1/templates/123/analyze/1234 HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.64.1
> Accept: */*
> Content-Type: application/json
> Content-Length: 57
>
* upload completely sent off: 57 out of 57 bytes
* Empty reply from server
* Connection #0 to host localhost left intact
curl: (52) Empty reply from server
* Closing connection 0
```

File: README.MD (hunk 1 above)
Comment (eladiw) [reply README.MD]: I will rephrase. I meant you dont need to have an k8s cluster.

File: deployment/deploy-presidio.sh (hunk 2 above)
Comment (eladiw) [reply deployment/deploy-presidio.sh]: done

Comment (eladiw) [reply deployment/deploy-presidio.sh]: actually I reverted these changes. we can move to helm 3 on another pr

File: README.MD (hunk 4 above)
Comment (eladiw) [reply README.MD]: I will add prereqs script

Comment (itye-msft) [reply README.MD]: If Kind is already cloned it will re-clone it.
Perhaps first check if it's already there.
Running the script twice in a row will cause:
1. Cloning again KIND.
2. Create a namepsace again. It will fail because it exists.
3. ./deploy-helm.sh Must run only once. It will fail with erros.
4. ./deploy-helm.sh Is obsolete because Helm 3 does not use Tiller. I got an error there.

If there are assumptions about running this script only once - it should be stated, or change the script to allow it running multiple times

File: README.MD (hunk 1 above)
Comment (eladiw) [reply README.MD]: done

File: README.MD (hunk 4 above)
Comment (eladiw) [reply README.MD]: added comment about the one time.
i dont want to bloat this script with retry mechanism and cleanups. sounds overkill

**Comment by eladiw** [comment]
@ityemsft please try with different port
@balteravishay  @omri374  @itye-msft  @tamirkamara 

Updated script to include prerequisites.sh
create a vm:
cd deployment
sudo ./prerequisites.sh
sudo ./run-with-kind.sh

and port forward will be done automatically...

File: README.MD [hunk 5]
```diff
@@ -88,6 +88,37 @@ Presidio leverages:
...
+
+        ```sh
+        cd deployment/
+        ./prerequisites.sh
+        ```
+
+        depending on your environment, sudo might be needed
+
+2. Clone Presidio.
+3. Run
```
Comment (omri374) [inline README.MD]: Could you add a sentence on what this script does?

File: deployment/run-with-kind.sh [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+# This script should be ran ONCE. if for anyreason a retry is needed.
+# cleanup of the local environment is needed (kind delete cluster)
+
+# Clone KIND and build it (this way we eliminate the need to have GO installed)
+git clone https://github.com/kubernetes-sigs/kind.git
+cd kind
+make build
+
+# Create a KIND cluster into which we will deploy Presidio
```
Comment (omri374) [inline deployment/run-with-kind.sh]: Is it possible to add notifications on the step it's currently add? some verbosity into the process as it takes a few minutes to complete

Comment (eladiw) [reply deployment/run-with-kind.sh]: done

File: README.MD (hunk 5 above)
Comment (eladiw) [reply README.MD]: done

File: README.MD [hunk 7]
```diff
@@ -88,6 +88,37 @@ Presidio leverages:
 * [GRPC](https://grpc.io)
 * [re2](https://github.com/google/re2)
 
+## Give it a try - Run locally with KIND
```
[the remark this answers is not in the document]
Comment (eladiw) [reply README.MD]: Done

**Comment by eladiw** [comment]
@omri374. Updated according to your comments

File: README.MD [hunk 8]
```diff
@@ -88,6 +88,44 @@ Presidio leverages:
...
+
+        ```sh
+        cd deployment/
+        ./prerequisites.sh
+        ```
+
+        depending on your environment, sudo might be needed
+
+2. Clone Presidio.
+|
```
[the remark this answers is not in the document]
Comment (eladiw) [reply README.MD]: done

File: README.MD [hunk 9]
```diff
@@ -88,6 +88,44 @@ Presidio leverages:
 * [GRPC](https://grpc.io)
 * [re2](https://github.com/google/re2)
 
+## Give it a try
```
[the remark this answers is not in the document]
Comment (eladiw) [reply README.MD]: done
